src/libraries/Common/src/Internal/Cryptography/UniversalCryptoEncryptor.cs

```diff
@@ -23,81 +23,130 @@ public UniversalCryptoEncryptor(PaddingMode paddingMode, BasicSymmetricCipher ba
         {
         }
 
-        protected sealed override int UncheckedTransformBlock(byte[] inputBuffer, int inputOffset, int inputCount, byte[] outputBuffer, int outputOffset)
+        protected override int UncheckedTransformBlock(ReadOnlySpan<byte> inputBuffer, Span<byte> outputBuffer)
         {
-            return BasicSymmetricCipher.Transform(inputBuffer, inputOffset, inputCount, outputBuffer, outputOffset);
+            return BasicSymmetricCipher.Transform(inputBuffer, outputBuffer);
         }
 
-        protected sealed override unsafe byte[] UncheckedTransformFinalBlock(byte[] inputBuffer, int inputOffset, int inputCount)
+        protected override int UncheckedTransformFinalBlock(ReadOnlySpan<byte> inputBuffer, Span<byte> outputBuffer)
         {
-            byte[] paddedBlock = PadBlock(inputBuffer, inputOffset, inputCount);
+            int padWritten = PadBlock(inputBuffer, outputBuffer);
+            int transformWritten = BasicSymmetricCipher.TransformFinal(outputBuffer.Slice(0, padWritten), outputBuffer);
```

Comment:
There are now two different times that the output buffer gets written to.  This could break something like

```C#
byte[] buf = new byte[1024];
Span<byte> dest = buf;
Span<byte> input = buf.Slice(41);
Fill(input);
Encrypt(input, dest)
```

If the padding happened to do an overwrite of the input buffer.

This could either use MemoryMarshal.Overlaps to see if there's enough space for a block of padding without contamination then falling back to a rented buffer, or just use a rented buffer.

Always renting a buffer is certainly an easier code-flow.

Reply:
Yeah, there is only one caller of this right now which is `byte[] UncheckedTransformFinalBlock(...)` and it is always guaranteed to pass in the buffer of the right size. I had presumed that "Unchecked" in the name was sufficient to indicate "Hey, you need to make sure you are passing in the right size", but if you think this API needs to be able to stand on its own, I can add something there that make sure there is enough room for padding.

It would just seem redundant right now for the one caller to say "this is the exactly right size you need" and the callee to rent a buffer anyway.

Reply:
So long as input and output don't overlap, this is fine.  Just throw in something like

```C#
// The only caller of this method is the array-allocating overload, outputBuffer is
// always new memory, not a user-provided buffer.
Debug.Assert(!inputBuffer.Overlaps(outputBuffer));
```